src/wasm/wasm-type.cpp

```diff
@@ -689,40 +708,56 @@ template<typename Info> bool Store<Info>::isGlobalStore() {
 #endif
 
 template<typename Info>
-typename Info::type_t Store<Info>::canonicalize(const Info& info) {
+typename Info::type_t Store<Info>::insert(const Info& info) {
   typename Info::type_t canonical;
   if (info.getCanonical(canonical)) {
     return canonical;
   }
   std::lock_guard<std::recursive_mutex> lock(mutex);
-  auto indexIt = typeIDs.find(std::cref(info));
-  if (indexIt != typeIDs.end()) {
-    return typename Info::type_t(indexIt->second);
+  if (std::is_same<Info, TypeInfo>::value ||
+      typeSystem == TypeSystem::Equirecursive) {
```

Comment:
why is TypeInfo special here?

Reply:
Using nominal types should only change the behavior of the HeapType store, not the Type store. It should not be possible, for example, to have two distinct nullable reference types to the same HeapType, so the Type store needs to keep canonicalizing ust as it does in the equirecursive system.

Reply:
I see, thanks! That's what I was missing.